### ISSUE TITLE
accelerator.gather() at training time

### ISSUE THREAD

**Issue by DeepakSaini119** [huggingface/accelerate#76]
Can I use accelerator.gather() at training time? Would gradients be calculated properly? Basically my use case is something like below toy snippet. It seems that there is some issue with gradient flow in this scheme as my validation accuracy drops to 0.

```python
model, optimizer, train_loader = accelerator.prepare(model, optimizer, train_loader)
for i, data in enumerate(train_loader):
    model.zero_grad()
    
    a, b = model(data)
    b_all = accelerator.gather(b)
    c = f(a, b_all)
    loss = criterion(a, b, c)
    accelerator.backward(loss)
    optimizer.step()
```

**Comment by sgugger** [comment]
Ah no, the `gather` method does not propagate the history of the tensors (necessary for the backward pass), so this won't work if you need gradients for `b_all`.

**Comment by DeepakSaini119** [comment]
Ok, thanks! Do we have a workaround for this?

**Comment by sgugger** [comment]
Not that I know of. How would you it in PyTorch distributed?

**Comment by ddkalamk** [comment]
Here is an example of distributed all_gather that works with autograd.
https://github.com/facebookresearch/dlrm/blob/master/extend_distributed.py#L579

Essentially, we need to wrap it under autograd Function to enable backward.

**Comment by DeepakSaini119** [comment]
Thanks for the suggestion @ddkalamk. I tried implementing it. It does propagate the history but the gradients are not being calculated properly. Any oblivious bug?
```python
import torch
from torch.autograd import Function
from accelerate import Accelerator

class AllGather(Function):
    @staticmethod
    def forward(ctx, tensor):
        local_start = tensor.shape[0] * torch.distributed.get_rank()

        output_size = list(tensor.size())

        ctx.local_start = local_start
        ctx.local_length = tensor.shape[0]

        tensor = tensor.contiguous()

        out_len = tensor.shape[0] * torch.distributed.get_world_size()
        output_size[0] = out_len
        
        output = tensor.new_empty(output_size)
        gather_list = list(output.split([tensor.shape[0]] * torch.distributed.get_world_size(), dim=0))

        torch.distributed.all_gather(gather_list, tensor)

        return output

    @staticmethod
    def backward(ctx, grad_output):
        start = ctx.local_start
        length = ctx.local_length

        grad_input = grad_output.narrow(0, start, length)

        return (grad_input, None, None)

accelerator = Accelerator()
DEVICE = accelerator.device

tensor = torch.ones((16, 4), requires_grad=True)
tensor = tensor.to(DEVICE)
print(f"tensor grad_fn = {tensor.grad_fn}")

# internal implementation of accelerate.gather() -> in this case the tensor_all.grad_fn is None
# output_tensors = [torch.zeros_like(tensor) for _ in range(torch.distributed.get_world_size())]
# torch.distributed.all_gather(output_tensors, tensor)
# tensor_all = torch.cat(output_tensors, dim=0)

tensor_all = AllGather.apply(tensor)

if accelerator.is_main_process:
    print(tensor_all.shape)
    print(f"tensor_all grad_fn = {tensor_all.grad_fn}")
```

**Comment by ddkalamk** [comment]
@DeepakSaini119  I am not sure, but assuming you are using DDP, you may need to scale the the gradient as DDP calculates average insetad of just allreduce (it divides gradients by world_size after allreduce). In some cases, this can effectivly done by scaling learning rate.
